# SQL CARTO QUERY FUNCTION

# written for PUI2017_Python2 by Gokmen Dedemen

In [1]:
SQL_SOURCE = 'https://gd1097.carto.com/api/v2/sql?q='

import ast
import pandas as pd
try:
    # Python 3
    from urllib.parse import urlparse, urlencode
    from urllib.request import urlopen, Request
    from urllib.error import HTTPError
except ImportError:
    # Python 2
    from urlparse import urlparse
    from urllib import urlencode
    from urllib2 import urlopen, Request, HTTPError
    
try:
    # Python 2
    from cStringIO import StringIO as BytesIO
except ImportError:
    # Python 3
    from io import BytesIO


def queryCartoDB(query, format='CSV', source=SQL_SOURCE):
    '''queries carto datasets from a given carto account
    Arguments: 
    query - string: a valid sql query string
    format - outlut format  OPTIONAL (default CSV)
    source - a valid sql api endpoint OPTIONAL (default carto fb55 account)
    Returns:
    the return of the sql query AS A STRING
    NOTES:
    designed for the carto API, tested only with CSV return format'''
    
    data = urlencode({'format': format, 'q': query})
    data=data.encode("utf-8")
    try:
        response = urlopen(source,data)
    except HTTPError as e:
        raise ValueError('\n'.join(ast.literal_eval(e.readline())['error']))
    except Exception:
        raise
    return response.read()

# The part above is for making the code compatible to both Python 2 and Python 3 for extra credit part


In [2]:
test = '''
SELECT *
FROM citibike
'''

In [3]:
df = pd.read_csv(BytesIO(queryCartoDB(test)), sep=',')


In [4]:
df.head()

cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        5960       NaN                   NaN       1      1965.0   19217   
1        9806       NaN                   NaN       1      1971.0   17575   
2       13113       NaN                   NaN       1      1973.0   20313   
3           1       NaN                   NaN       2      1978.0   17131   
4           2       NaN                   NaN       1      1993.0   21289   

   end_station_id  start_station_id  tripduration    usertype  \
0             462               435           351  Subscriber   
1             488               167           643  Subscriber   
2             440               332          1379  Subscriber   
3             423               521           801  Subscriber   
4             504               497           379  Subscriber   

   end_station_name  start_station_name  end_station_longitude  \
0  W 22 St & 10 Ave     W 21 St & 6 Ave             -74.004519   
1   W 39 St & 9 Ave     E 39 St & 3 Ave             -73.993722   
2   E 45 St & 3 Ave           Cherry St             -73.972826   
3   W 54 St & 9 Ave     8 Ave & W 31 St             -73.986905   
4   1 Ave & E 15 St  E 17 St & Broadway             -73.981656   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.746920               -73.994156               40.741740   
1             40.756458               -73.976049               40.748901   
2             40.752554               -73.979481               40.712199   
3             40.765849               -73.994811               40.750450   
4             40.732219               -73.990093               40.737050   

                 stoptime               starttime  
0  2015-02-01 20:42:00+00  2015-02-01 20:36:00+00  
1  2015-02-03 14:50:00+00  2015-02-03 14:39:00+00  
2  2015-02-04 07:30:00+00  2015-02-04 07:07:00+00  
3  2015-02-01 00:14:00+00  2015-02-01 00:00:00+00  
4  2015-02-01 00:07:00+00  2015-02-01 00:00:00+00

# TASK 1
## a) Sort data by start_station_id, tripduration
### b) Only checking trips with duration <= 3 hours
### c) Only show the top/last 10 records (aka head and tail in SQL) List all unique start_station_id values
### d) Aggregation functions:
### Count the number of trips (aka wc -l in SQL) Find the average/min/max trip duration

# SOLUTION
## a) let's first sort all data by start_station_id, tripduration

In [5]:
task = '''
SELECT * 
FROM citibike
ORDER BY start_station_id, tripduration
'''

In [6]:
pd.read_csv(BytesIO(queryCartoDB(task)), sep=',')


cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0            8998       NaN                   NaN       2      1980.0   19039   
1           40759       NaN                   NaN       2      1959.0   18131   
2           40757       NaN                   NaN       1      1995.0   19729   
3           21948       NaN                   NaN       1      1970.0   16879   
4           43099       NaN                   NaN       1      1983.0   15277   
5           32566       NaN                   NaN       1      1971.0   18492   
6           15971       NaN                   NaN       1      1961.0   19216   
7           41355       NaN                   NaN       1      1974.0   21588   
8           42074       NaN                   NaN       1      1978.0   19124   
9            2282       NaN                   NaN       1      1974.0   18174   
10          44674       NaN                   NaN       1      1983.0   18198   
11           2142       NaN                   NaN       1      1982.0   19672   
12            649       NaN                   NaN       2      1986.0   14692   
13          44789       NaN                   NaN       1      1983.0   17669   
14           9494       NaN                   NaN       1      1961.0   16754   
15          42446       NaN                   NaN       1      1983.0   21114   
16          41465       NaN                   NaN       1      1970.0   19729   
17          14395       NaN                   NaN       1      1981.0   18620   
18          19984       NaN                   NaN       1      1981.0   20119   
19           8726       NaN                   NaN       1      1981.0   20988   
20          26155       NaN                   NaN       1      1978.0   18287   
21          25842       NaN                   NaN       2      1972.0   16310   
22          25587       NaN                   NaN       1      1973.0   18991   
23          24156       NaN                   NaN       1      1981.0   20600   
24          27104       NaN                   NaN       1      1973.0   15927   
25           3957       NaN                   NaN       1      1991.0   15282   
26          31639       NaN                   NaN       1      1978.0   16405   
27           9194       NaN                   NaN       1      1980.0   17196   
28          33221       NaN                   NaN       2      1986.0   17972   
29          38564       NaN                   NaN       1      1974.0   19771   
...           ...       ...                   ...     ...         ...     ...   
46170       31410       NaN                   NaN       1      1958.0   17595   
46171       22533       NaN                   NaN       1      1962.0   15178   
46172        1333       NaN                   NaN       0         NaN   18288   
46173       23807       NaN                   NaN       1      1974.0   16147   
46174       27637       NaN                   NaN       1      1973.0   17061   
46175       22399       NaN                   NaN       1      1991.0   17198   
46176       24714       NaN                   NaN       1      1958.0   20872   
46177       41541       NaN                   NaN       1      1955.0   15495   
46178       43007       NaN                   NaN       1      1966.0   20780   
46179       35349       NaN                   NaN       1      1956.0   17077   
46180       20877       NaN                   NaN       1      1986.0   20268   
46181       39913       NaN                   NaN       1      1954.0   15586   
46182       38954       NaN                   NaN       1      1986.0   19331   
46183       30026       NaN                   NaN       1      1949.0   15627   
46184       14538       NaN                   NaN       1      1959.0   17488   
46185        2368       NaN                   NaN       0         NaN   16918   
46186       24145       NaN                   NaN       1      1959.0   15102   
46187       32820       NaN           

# b) Only checking trips with duration <= 3 hours
# c) Only show the top/last 10 records (aka head and tail in SQL) List all unique start_station_id values

# FIRST TOP 10 RECORDS

In [7]:
task2 = '''
SELECT * 
FROM citibike
WHERE tripduration <= 10800
ORDER BY start_station_id, tripduration
LIMIT 10
'''

In [8]:
pd.read_csv(BytesIO(queryCartoDB(task2)), sep=',')



cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        8998       NaN                   NaN       2        1980   19039   
1       40759       NaN                   NaN       2        1959   18131   
2       40757       NaN                   NaN       1        1995   19729   
3       21948       NaN                   NaN       1        1970   16879   
4       43099       NaN                   NaN       1        1983   15277   
5       32566       NaN                   NaN       1        1971   18492   
6       15971       NaN                   NaN       1        1961   19216   
7       41355       NaN                   NaN       1        1974   21588   
8       42074       NaN                   NaN       1        1978   19124   
9        2282       NaN                   NaN       1        1974   18174   

   end_station_id  start_station_id  tripduration    usertype  \
0              72                72           107  Subscriber   
1              72                72           189  Subscriber   
2              72                72           229  Subscriber   
3             478                72           248  Subscriber   
4             515                72           252  Subscriber   
5             478                72           261  Subscriber   
6             478                72           276  Subscriber   
7             447                72           283  Subscriber   
8             514                72           288  Subscriber   
9             447                72           301  Subscriber   

   end_station_name start_station_name  end_station_longitude  \
0  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
1  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
2  W 52 St & 11 Ave   W 52 St & 11 Ave             -73.993929   
3  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
4  W 43 St & 10 Ave   W 52 St & 11 Ave             -73.994618   
5  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
6  11 Ave & W 41 St   W 52 St & 11 Ave             -73.998842   
7   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   
8  12 Ave & W 40 St   W 52 St & 11 Ave             -74.002777   
9   8 Ave & W 52 St   W 52 St & 11 Ave             -73.985162   

   end_station_latitude  start_station_longitude  start_station_latitude  \
0             40.767272               -73.993929               40.767272   
1             40.767272               -73.993929               40.767272   
2             40.767272               -73.993929               40.767272   
3             40.760301               -73.993929               40.767272   
4             40.760094               -73.993929               40.767272   
5             40.760301               -73.993929               40.767272   
6             40.760301               -73.993929               40.767272   
7             40.763707               -73.993929               40.767272   
8             40.760875               -73.993929               40.767272   
9             40.763707               -73.993929               40.767272   

                 stoptime               starttime  
0  2015-02-03 10:12:00+00  2015-02-03 10:10:00+00  
1  2015-02-07 10:43:00+00  2015-02-07 10:40:00+00  
2  2015-02-07 10:44:00+00  2015-02-07 10:40:00+00  
3  2015-02-05 06:24:00+00  2015-02-05 06:19:00+00  
4  2015-02-07 15:47:00+00  2015-02-07 15:43:00+00  
5  2015-02-06 09:02:00+00  2015-02-06 08:58:00+00  
6  2015-02-04 12:51:00+00  2015-02-04 12:46:00+00  
7  2015-02-07 12:15:00+00  2015-02-07 12:10:00+00  
8  2015-02-07 13:54:00+00  2015-02-07 13:50:00+00  
9  2015-02-01 13:13:00+00  2015-02-01 13:08:00+00

## THEN, LAST 10 RECORDS

In [9]:
task3 = '''
SELECT * 
FROM citibike
WHERE tripduration <= 10800
ORDER BY (start_station_id, tripduration) DESC
LIMIT 10
'''

In [10]:
pd.read_csv(BytesIO(queryCartoDB(task3)), sep=',')



cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0        1506       NaN                   NaN       0         NaN   21202   
1        8831       NaN                   NaN       1      1958.0   15552   
2       23529       NaN                   NaN       1      1985.0   21473   
3       10173       NaN                   NaN       1      1969.0   15309   
4       27203       NaN                   NaN       1      1969.0   20069   
5        6781       NaN                   NaN       1      1959.0   19665   
6       44703       NaN                   NaN       2      1964.0   15443   
7       11587       NaN                   NaN       1      1996.0   19133   
8        1228       NaN                   NaN       0         NaN   17849   
9       17590       NaN                   NaN       1      1969.0   20246   

   end_station_id  start_station_id  tripduration    usertype  \
0             499              3002          7225    Customer   
1             510              3002          2139  Subscriber   
2             359              3002          2131  Subscriber   
3             318              3002          1977  Subscriber   
4             318              3002          1977  Subscriber   
5             494              3002          1976  Subscriber   
6             410              3002          1920  Subscriber   
7             279              3002          1908  Subscriber   
8             525              3002          1900    Customer   
9             318              3002          1830  Subscriber   

           end_station_name          start_station_name  \
0        Broadway & W 60 St  South End Ave & Liberty St   
1           W 51 St & 6 Ave  South End Ave & Liberty St   
2        E 47 St & Park Ave  South End Ave & Liberty St   
3  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   
4  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   
5           W 26 St & 8 Ave  South End Ave & Liberty St   
6   Suffolk St & Stanton St  South End Ave & Liberty St   
7      Peck Slip & Front St  South End Ave & Liberty St   
8          W 34 St & 11 Ave  South End Ave & Liberty St   
9  E 43 St & Vanderbilt Ave  South End Ave & Liberty St   

   end_station_longitude  end_station_latitude  start_station_longitude  \
0             -73.981918             40.769155               -74.015756   
1             -73.980420             40.760660               -74.015756   
2             -73.974987             40.755103               -74.015756   
3             -73.977987             40.753202               -74.015756   
4             -73.977987             40.753202               -74.015756   
5             -73.997236             40.747348               -74.015756   
6             -73.985180             40.720664               -74.015756   
7             -74.001670             40.707873               -74.015756   
8             -74.002116             40.755942               -74.015756   
9             -73.977987             40.753202               -74.015756   

   start_station_latitude                stoptime               starttime  
0               40.711512  2015-02-01 13:47:00+00  2015-02-01 11:46:00+00  
1               40.711512  2015-02-03 10:09:00+00  2015-02-03 09:33:00+00  
2               40.711512  2015-02-05 09:17:00+00  2015-02-05 08:42:00+00  
3               40.711512  2015-02-03 16:37:00+00  2015-02-03 16:04:00+00  
4               40.711512  2015-02-05 16:44:00+00  2015-02-05 16:11:00+00  
5               40.711512  2015-02-02 09:53:00+00  2015-02-02 09:21:00+00  
6               40.711512  2015-02-07 19:08:00+00  2015-02-07 18:36:00+00  
7               40.711512  2015-02-03 19:26:00+00  2015-02-03 18:54:00+00  
8               40.711512  2015-02-01 11:45:00+00  2015-02-01 11:14:00+00  
9               40.711512  2015-02-04 17:04:00+00  2015-02-04 16:33:00+00

# List all unique start_station_id values

In [11]:
df.start_station_id.unique()

array([ 435,  167,  332,  521,  497,  281, 2004,  323,  373,  352,  439,
        335,  284,  498,  368,  326,  285,  394,  493,  445,  428,  228,
        513,  507,  128,  490,  350,  250,  504,  317,  461,  477,  531,
        300,  223,  472,  173,  487,  444,  515,  327,  243,  328,  168,
        312,  254,  539,  465,  383,  237,  447,  290,  434,  345,  511,
        449,  334,  119,  453,  438,  293,  150,  195,  401,  274,  271,
        151,  279,  403,  212, 2022,  503,  405,  491,  276,  325,  486,
        324,  417,  291,  479,  348,  485, 2021,  404,  432,  236,  164,
        499,  478,  501,  433,  238,  336, 2002,  157,  267,  528,  505,
        382,  489,  537,  411,  161,  483,  437,  393,  482,  520,  252,
        251,  307,  470, 2012,  518,  533,  369,  174,  488,  476,  523,
        379,  258,  421,  462,  301,  377,  436,  494,  492,  143,  295,
        514,  527,  448,  349,  530,  468,  366,  153,   79,  340,  500,
        502, 3002,  391,  469,  458,  116,  415,  3

## Aggregation functions:
## Count the number of trips (aka wc -l in SQL) Find the average/min/max trip duration

In [12]:
task4 = '''
SELECT
COUNT(tripduration),
AVG(tripduration),
MIN(tripduration),
MAX(tripduration)

FROM citibike
'''


In [13]:
pd.read_csv(BytesIO(queryCartoDB(task4)), sep=',')



count         avg  min    max
0  46200  675.865823   60  43016

# Task 2 — Working with date/time
## a) Selecting trips started on Feb-02-2015 only

## b) Selecting trips started on the weekends

## c) What are average trip duration during weekends? Can we do the same for weekday?

# SOLUTION
## a) Selecting trips started on Feb-02-2015 only

In [14]:
task5 = '''
SELECT * FROM citibike
WHERE starttime >= '2015-02-02 00:00'
AND starttime < '2015-02-03 00:00'
'''

In [15]:
pd.read_csv(BytesIO(queryCartoDB(task5)), sep=',')



cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0           6442       NaN                   NaN       1      1992.0   20684   
1           6443       NaN                   NaN       2      1964.0   16094   
2           6444       NaN                   NaN       1      1974.0   17362   
3           6445       NaN                   NaN       2      1992.0   15475   
4           6446       NaN                   NaN       1      1968.0   16272   
5           6447       NaN                   NaN       1      1979.0   16370   
6           6448       NaN                   NaN       1      1963.0   19437   
7           6449       NaN                   NaN       1      1982.0   21211   
8           6450       NaN                   NaN       2      1978.0   21417   
9           6451       NaN                   NaN       1      1956.0   15722   
10          6452       NaN                   NaN       1      1962.0   18972   
11          6453       NaN                   NaN       1      1980.0   19628   
12          6454       NaN                   NaN       1      1978.0   15169   
13          6455       NaN                   NaN       1      1997.0   17533   
14          6456       NaN                   NaN       1      1983.0   19761   
15          6457       NaN                   NaN       1      1993.0   17267   
16          6458       NaN                   NaN       1      1964.0   18691   
17          6459       NaN                   NaN       1      1977.0   17106   
18          6460       NaN                   NaN       1      1973.0   20009   
19          6461       NaN                   NaN       1      1979.0   18777   
20          6462       NaN                   NaN       1      1989.0   15357   
21          6463       NaN                   NaN       1      1983.0   18487   
22          6464       NaN                   NaN       1      1981.0   20863   
23          6465       NaN                   NaN       1      1970.0   16539   
24          6466       NaN                   NaN       1      1961.0   18737   
25          6467       NaN                   NaN       2      1980.0   17797   
26          6468       NaN                   NaN       1      1990.0   20638   
27          6469       NaN                   NaN       1      1989.0   17636   
28          6470       NaN                   NaN       1      1982.0   18472   
29          6471       NaN                   NaN       1      1968.0   17456   
...          ...       ...                   ...     ...         ...     ...   
1429        7871       NaN                   NaN       1      1985.0   20572   
1430        7872       NaN                   NaN       1      1981.0   20338   
1431        7873       NaN                   NaN       1      1978.0   15826   
1432        7874       NaN                   NaN       1      1990.0   18325   
1433        7875       NaN                   NaN       1      1977.0   16654   
1434        7876       NaN                   NaN       1      1966.0   20902   
1435        7877       NaN                   NaN       1      1954.0   20796   
1436        7878       NaN                   NaN       1      1989.0   20160   
1437        7879       NaN                   NaN       1      1974.0   17392   
1438        7880       NaN                   NaN       1      1963.0   17263   
1439        7881       NaN                   NaN       1      1955.0   19858   
1440        7882       NaN                   NaN       1      1979.0   20319   
1441        7883       NaN                   NaN       1      1977.0   21600   
1442        7884       NaN                   NaN       1      1986.0   18791   
1443        7885       NaN                   NaN       1      1966.0   17730   
1444        7886       NaN                   NaN       1      1980.0   18147   
1445        7887       NaN                   NaN       1      1959.0   20249   
1446        7888       NaN                   NaN       1      1988.0   18078   
1447  

## b)  Selecting trips started on the weekends



In [16]:
task6 = '''
SELECT *
FROM citibike
WHERE extract (DOW from starttime) IN (0,6)
'''

In [17]:
pd.read_csv(BytesIO(queryCartoDB(task6)), sep=',')



cartodb_id  the_geom  the_geom_webmercator  gender  birth_year  bikeid  \
0            5960       NaN                   NaN       1      1965.0   19217   
1               1       NaN                   NaN       2      1978.0   17131   
2               2       NaN                   NaN       1      1993.0   21289   
3               3       NaN                   NaN       2      1969.0   18903   
4               4       NaN                   NaN       2      1985.0   21044   
5               5       NaN                   NaN       1      1957.0   19868   
6               6       NaN                   NaN       1      1979.0   15854   
7               7       NaN                   NaN       1      1983.0   15173   
8               8       NaN                   NaN       1      1955.0   17862   
9               9       NaN                   NaN       2      1985.0   21183   
10             10       NaN                   NaN       1      1982.0   14843   
11             11       NaN                   NaN       2      1988.0   16936   
12             12       NaN                   NaN       2      1991.0   21507   
13             13       NaN                   NaN       2      1982.0   17456   
14             14       NaN                   NaN       2      1990.0   14679   
15             15       NaN                   NaN       1      1977.0   18978   
16             16       NaN                   NaN       1      1968.0   16496   
17             17       NaN                   NaN       1      1990.0   17107   
18             18       NaN                   NaN       1      1987.0   16984   
19             19       NaN                   NaN       1      1980.0   15997   
20             20       NaN                   NaN       1      1960.0   21516   
21             21       NaN                   NaN       1      1986.0   15815   
22             22       NaN                   NaN       1      1987.0   18727   
23             23       NaN                   NaN       1      1969.0   14982   
24             24       NaN                   NaN       1      1957.0   16391   
25             25       NaN                   NaN       1      1981.0   15947   
26             26       NaN                   NaN       1      1993.0   16917   
27             27       NaN                   NaN       1      1975.0   15924   
28             28       NaN                   NaN       1      1973.0   21321   
29             29       NaN                   NaN       2      1973.0   16861   
...           ...       ...                   ...     ...         ...     ...   
13201       46171       NaN                   NaN       1      1988.0   15733   
13202       46172       NaN                   NaN       1      1995.0   16931   
13203       46173       NaN                   NaN       1      1987.0   17202   
13204       46174       NaN                   NaN       2      1989.0   20314   
13205       46175       NaN                   NaN       1      1989.0   21531   
13206       46176       NaN                   NaN       1      1968.0   21491   
13207       46177       NaN                   NaN       1      1990.0   15367   
13208       46178       NaN                   NaN       2      1973.0   18978   
13209       46179       NaN                   NaN       1      1980.0   19507   
13210       46180       NaN                   NaN       1      1980.0   16936   
13211       46181       NaN                   NaN       1      1982.0   18689   
13212       46182       NaN                   NaN       1      1972.0   15781   
13213       46183       NaN                   NaN       1      1988.0   15081   
13214       46184       NaN                   NaN       1      1983.0   21672   
13215       46185       NaN                   NaN       1      1970.0   16336   
13216       46186       NaN                   NaN       1      1973.0   21609   
13217       46187       NaN                   NaN       1      1964.0   19470   
13218       46188       NaN           

## c) What are average trip duration during weekends? Can we do the same for weekday?

### NOW FOR THE WEEKENDS

In [18]:
task7 = '''
SELECT AVG(tripduration) as avg
FROM citibike
WHERE extract (DOW from starttime) IN (0,6)
'''

In [19]:
pd.read_csv(BytesIO(queryCartoDB(task7)), sep=',')



avg  Unnamed: 1
0  662.942181         NaN

## For the weekends, the average trip duration is 662.94 seconds as seen above

### THEN FOR THE WEEKDAYS

In [20]:
task8 = '''
SELECT AVG(tripduration)
FROM citibike
WHERE extract (DOW from starttime) IN (0,1,2,3,4,5)
'''

In [21]:
pd.read_csv(BytesIO(queryCartoDB(task8)), sep=',')



avg  Unnamed: 1
0  674.040396         NaN

## For the weekdays, the average trip duration is 674.04 seconds as seen above

# Task 3 — Working with Space
## a) Showing the list of start station locations Using GROUP BY

## b) Showing the number of trips started per station 

## c) Showing the number of trips started per station but only for stations within 500m of Time Square!
#### NOTE: The coordinates of Time Square is (40.7577,-73.9857)

# SOLUTION
## a) Showing the list of start station locations Using GROUP BY 
## &
## b) Showing the number of trips started per station



Assuming that by mentioning start station locations, the question asks latitude, longitude, and start_station_id information.


Here I am using group by for start station locations (id, latitude and longitude) and I am also counting the number of trips from that location

In [22]:
task9='''
SELECT start_station_id, 
       start_station_latitude, 
       start_station_longitude, 
COUNT(*) FROM citibike
GROUP BY start_station_id, 
         start_station_latitude, 
         start_station_longitude

'''

In [23]:
pd.read_csv(BytesIO(queryCartoDB(task9)), sep=',')



start_station_id  start_station_latitude  start_station_longitude  count
0                 438               40.727791               -73.985649    156
1                 342               40.717400               -73.980166     78
2                 242               40.697883               -73.973503     46
3                 161               40.729170               -73.998102    203
4                 373               40.693317               -73.953820     27
5                 168               40.739713               -73.994564    327
6                 502               40.714215               -73.981346    151
7                 421               40.695734               -73.971297     37
8                 511               40.729387               -73.977724    279
9                 422               40.770513               -73.988038    100
10                150               40.720874               -73.980858    111
11                346               40.736529               -74.006180    145
12                336               40.730477               -73.999061    168
13                447               40.763707               -73.985162    133
14                224               40.711464               -74.005524     85
15                456               40.759711               -73.974023    110
16                233               40.692463               -73.989639     78
17                391               40.697601               -73.993446     69
18                329               40.720434               -74.010206    145
19                403               40.725029               -73.990697    151
20                497               40.737050               -73.990093    455
21                473               40.721101               -73.991925     99
22               2005               40.705312               -73.971001      7
23                238               40.736197               -74.008592    125
24                315               40.703554               -74.006702     69
25                312               40.722055               -73.989111    207
26                368               40.730386               -74.002150    238
27                496               40.737262               -73.992390    250
28                195               40.709056               -74.010434    137
29                454               40.754557               -73.965930     55
..                ...                     ...                      ...    ...
297               366               40.693261               -73.968896     88
298               433               40.729554               -73.980572    259
299               348               40.724910               -74.001547    159
300               475               40.735243               -73.987586    156
301               404               40.740583               -74.005509    111
302               212               40.743349               -74.006818    186
303               261               40.694749               -73.983625     50
304               530               40.771522               -73.990541    120
305               152               40.714740               -74.009106     71
306               393               40.722992               -73.979955    142
307               257               40.719392               -74.002472    123
308               279               40.707873               -74.001670     89
309               344               40.685144               -73.953809     34
310               449               40.764618               -73.987895    107
311               358               40.732916               -74.007114    292
312               464               40.759345               -73.967597     41
313               390               40.692216               -73.984284     27
314               458               40.751396               -74.005226    172
315               345               40.736494               -73.997044    257
316               431      

## c) Showing the number of trips started per station but only for stations within 500m of Time Square!



In [24]:
task10='''

SELECT start_station_id, 
       start_station_latitude, 
       start_station_longitude, 
       COUNT(*) 
FROM citibike
WHERE ST_DWithin(CDB_LatLng(start_station_latitude,
                            start_station_longitude)::geography,
                            CDB_LatLng(40.7577, -73.9857)::geography, 500)                            
GROUP BY start_station_id, 
         start_station_latitude, 
         start_station_longitude

'''



In [25]:
pd.read_csv(BytesIO(queryCartoDB(task10)), sep=',')


start_station_id  start_station_latitude  start_station_longitude  count
0               477               40.756405               -73.990026    507
1              2021               40.759291               -73.988597    141
2               465               40.755136               -73.986580    251
3               493               40.756800               -73.982912    141
4               524               40.755273               -73.983169    112
5               529               40.757570               -73.990985    221
6               173               40.760647               -73.984427    213

# Task 4 — Putting it all together
## Find the station that had the longest average trip duration during weekends and within 500m of TimeSquare!

In [26]:
task11='''

SELECT start_station_id, 
       start_station_latitude, 
       start_station_longitude,
       AVG(tripduration)

FROM citibike
WHERE extract (DOW from starttime) IN (0,6) AND
ST_DWithin(CDB_LatLng(start_station_latitude,
                      start_station_longitude)::geography,
                      CDB_LatLng(40.7577, -73.9857)::geography, 500)  
                            
GROUP BY start_station_id, 
         start_station_latitude, 
         start_station_longitude
         
ORDER BY avg DESC
LIMIT 1
'''



In [27]:
pd.read_csv(BytesIO(queryCartoDB(task11)), sep=',')



start_station_id  start_station_latitude  start_station_longitude  \
0               173               40.760647               -73.984427   

           avg  
0  1010.104167

#### The station that had the longest average trip duration during weekends and within 500 m of Timesquare is station 173 with an average trip duration of 1010.10 seconds

# Extra: create lines for trips started from stations within 500m of Times Squares and lasted less than 2 hours. The number of trips per each pair of stations are output as attributes of these lines.

In [28]:
task12='''

SELECT start_station_id,
       end_station_id,
       ST_Makeline(CDB_TransformToWebmercator(
       CDB_LatLng(start_station_latitude,
                  start_station_longitude
       )),
         CDB_TransformToWebmercator(CDB_LatLng(
              end_station_latitude,
              end_station_longitude))
    ) as the_geom_webmercator, 
       start_station_latitude, 
       start_station_longitude,
       COUNT(*)


FROM citibike
WHERE tripduration < 7200 AND
ST_DWithin(CDB_LatLng(start_station_latitude,
                      start_station_longitude)::geography,
                      CDB_LatLng(40.7577, -73.9857)::geography, 500)  
                            
GROUP BY start_station_id, 
         start_station_latitude, 
         start_station_longitude,
         end_station_id,
         end_station_latitude,
         end_station_longitude
         
'''


In [29]:
pd.read_csv(BytesIO(queryCartoDB(task12)), sep=',')


start_station_id  end_station_id  \
0                 173             528   
1                 477             435   
2                 465             538   
3                 493             512   
4                 524              72   
5                 465             446   
6                 524             160   
7                2021             459   
8                 493             442   
9                 477             263   
10                465             332   
11                173             493   
12                493             454   
13               2021             495   
14                529             164   
15                493             477   
16               2021             153   
17                465             254   
18                529             422   
19                477             284   
20                477             318   
21                524             515   
22                493             478   
23                477             465   
24                477             492   
25                465             444   
26                477             500   
27               2021             510   
28                477             489   
29                477             486   
..                ...             ...   
542               529             367   
543               173             508   
544               173             537   
545               493             515   
546               173            3002   
547               524             478   
548               477             152   
549               465             496   
550               493             457   
551              2021            2023   
552               465             527   
553               493            2021   
554               465             461   
555               529             493   
556               477             294   
557               524             442   
558               493             511   
559               493             160   
560               173             164   
561               477             415   
562               477             450   
563               529             466   
564               477             530   
565               524             512   
566              2021             520   
567               173             422   
568              2021             490   
569               529             362   
570               524             477   
571               465             280   

                                  the_geom_webmercator  \
0    0102000020110F00000200000020D6742CE16A5FC112F4...   
1    0102000020110F000002000000F55695027D6B5FC186E5...   
2    0102000020110F00000200000004AF971C1D6B5FC17A48...   
3    0102000020110F0000020000009BB87D02B76A5FC146C1...   
4    0102000020110F000002000000F22A632FBE6A5FC1363A...   
5    0102000020110F00000200000004AF971C1D6B5FC17A48...   
6    0102000020110F000002000000F22A632FBE6A5FC1363A...   
7    0102000020110F0000020000001607D538556B5FC119A6...   
8    0102000020110F0000020000009BB87D02B76A5FC146C1...   
9    0102000020110F000002000000F55695027D6B5FC186E5...   
10   0102000020110F00000200000004AF971C1D6B5FC17A48...   
11   0102000020110F00000200000020D6742CE16A5FC112F4...   
12   0102000020110F0000020000009BB87D02B76A5FC146C1...   
13   0102000020110F0000020000001607D538556B5FC119A6...   
14   0102000020110F0000020000003F9300B2976B5FC1B8B8...   
15   0102000020110F0000020000009BB87D02B76A5FC146C1...   
16   0102000020110F0000020000001607D538556B5FC119A6...   
17   0102000020110F00000200000004AF971C1D6B5FC17A48...   
18   0102000020110F0000020000003F9300B2976B5FC1B8B8...   
19   0102000020110F000002000000F55695027D6B5FC186E5...   
20   0102000020110F000002000000F55695027D6B5FC186E5...   
21   0102000020110F000002000000F22A632FBE6A5FC1363A...   
22   0102000020110F0000020000009BB87D02B76A5FC146C1...   
23   0102000020110F000002000000F55695027D6B5FC186E5...   
24   0102000

# The column "count" ( the last column) denotes attributes of the lines in the dataframe above.